pkg/gen/gen.bzl

```diff
@@ -20,82 +84,184 @@ def _go_proto_srcs_impl(ctx):
     generated_files[prefix] = [f for f in srcs.srcs]
 
   return [
-    GeneratedFileInfo(
+    _GeneratedFileInfo(
       generated_files = generated_files,
       # Create a task to remove any existing protobuf files.
       cleanup_tasks = [
-# Use a subshell with () to avoid changing the directory in the main shell.
-"""
-(
-    cd "${BUILD_WORKSPACE_DIRECTORY}"
-    # Avoid searching the node_modules directory because it's full of
-    # irrelevant files.
-    find ./pkg -name node_modules -prune -o \
-        -type f -name '*.pb.go' -exec rm {} + -o \
-        -type f -name '*.pb.gw.go' -exec rm {} +
-)
-""",
+        _subshell_in_workspace_snippet([
+          _find_relevant + " -type f -name {} -exec rm {{}} +".format(suffix)
+        ]) for suffix in ["*.pb.go", "*.pb.gw.go"]
       ],
     )
   ]
 
-go_proto_srcs = rule(
-   implementation = _go_proto_srcs_impl,
-   attrs = {
-       "_srcs": attr.label_list(providers = [GoSource], default=PROTOBUF_SRCS),
-   },
+_go_proto_srcs = rule(
+  implementation = _go_proto_srcs_impl,
+  attrs = {
+   "_srcs": attr.label_list(providers = [GoSource], default=PROTOBUF_SRCS),
+  },
 )
 
-_cp_file_cmd_fmt = """
+# This rule is the default rule to build construct the input to _hoist_files
+# for srcs which have a path in the sandbox that is parallel to where those
+# files should end up in the repo.
+def _no_prefix_impl(ctx):
+  files = [f for di in ctx.attr.srcs for f in di[DefaultInfo].files.to_list()]
```

Comment:
Out of curiosity: how'd you work through the starlark API here? Is there an IDE like thing? Looking at docs? Using python reflection to figure out what fields/methods are available + printf-ing your way through?